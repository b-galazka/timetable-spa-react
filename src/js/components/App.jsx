import React, { Component } from 'react';
import { Switch, Route, BrowserRouter, Redirect } from 'react-router-dom';

import AppContainer from './AppContainer';
import NotFound from './NotFound';
import AppMobile from './AppMobile';
import detectAndroidDevice from '../utils/detectAndroidDevice';

import urlTranslations from '../../json/urlTranslations';
import urls from '../../json/urls.json';

import '../../scss/main.scss';

export default class App extends Component {

    constructor() {

        super();

        this.state = {
            android: detectAndroidDevice()
        };

        this.visitPage = this.visitPage.bind(this);
    }

    render() {

        return (
            <BrowserRouter>

                {
                    (this.state.android) ?

                        <AppMobile visitPage={this.visitPage} /> :

                        <main className="app">
                            <Switch>
                                <Redirect exact from="/" to={urls.defaultRedirection} />
                                <Route exact path='/:type/:slug' render={App.isUrlValid} />
                                <Route component={NotFound} />
                            </Switch>
                        </main>
                }

            </BrowserRouter>
        );
    }

    static isUrlValid({ match }) {

        const { type } = match.params;
        const possibleTypes = urlTranslations.types;

        if (possibleTypes.includes(type)) {

            return <AppContainer />;
        }

        return <NotFound />;
    }

    visitPage() {

        this.setState({
            android: false
        });
    }
}