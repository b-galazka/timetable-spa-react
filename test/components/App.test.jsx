import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import App from '../../src/js/components/App';

jest.mock('../../src/js/components/AppMobile', () => 'AppMobile');
jest.mock('../../src/js/components/NotFound', () => 'NotFound');
jest.mock('../../src/json/urlTranslations', () => ({ types: ['a'] }));
jest.mock('../../src/json/urls', () => ({ defaultRedirection: '/redirected' }));
jest.mock('../../src/js/components/AppContainer', () => 'AppContainer');

describe('App component', () => {

    it('should render mobile version if app is running on android device', () => {

        const { userAgent } = navigator;

        Object.defineProperty(navigator, 'userAgent', {
            value: 'lorem Android ipsum',
            configurable: true
        });

        const tree = renderer.create(<App />).toJSON();

        expect(tree).toMatchSnapshot();

        Object.defineProperty(navigator, 'userAgent', {
            value: userAgent,
            configurable: true
        });
    });

    it('should render notFound page if URL is invalid', () => {

        const tree = renderer.create(
            <MemoryRouter initialEntries={['/unknown-url']}>
                <App />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render notFound page if timetable object type is invalid', () => {

        const tree = renderer.create(
            <MemoryRouter initialEntries={['/b/c']}>
                <App />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render AppContiner if url is valid', () => {

        const tree = renderer.create(
            <MemoryRouter initialEntries={['/a/c']}>
                <App />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should redirect from "/" to URL set in config file', () => {

        const tree = renderer.create(
            <MemoryRouter initialEntries={['/']}>
                <App />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});