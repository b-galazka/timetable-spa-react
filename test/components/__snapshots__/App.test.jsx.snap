// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should redirect from "/" to URL set in config file 1`] = `
<div
  data-mock="BrowserRouter"
>
  <main
    className="app"
  >
    <div
      data-from="/"
      data-mock="Redirect"
      data-to="/redirected"
    />
  </main>
</div>
`;

exports[`App component should render AppContiner if url is valid 1`] = `
<div
  data-mock="BrowserRouter"
>
  <main
    className="app"
  >
    <AppContainer />
  </main>
</div>
`;

exports[`App component should render mobile version if app is running on android device 1`] = `
<div
  data-mock="BrowserRouter"
>
  <AppMobile
    visitPage={[Function]}
  />
</div>
`;

exports[`App component should render notFound page if URL is invalid 1`] = `
<div
  data-mock="BrowserRouter"
>
  <main
    className="app"
  >
    <NotFound />
  </main>
</div>
`;

exports[`App component should render notFound page if timetable object type is invalid 1`] = `
<div
  data-mock="BrowserRouter"
>
  <main
    className="app"
  >
    <NotFound />
  </main>
</div>
`;
